server/commands/list_objects.go

```diff
@@ -31,15 +34,100 @@ type ListObjectsQuery struct {
 	ListObjectsDeadline   time.Duration
 	ListObjectsMaxResults uint32
 	ResolveNodeLimit      uint32
+	ConnectedObjects      func(ctx context.Context, req *ConnectedObjectsRequest, results chan<- string) error
 }
 
-// Execute the ListObjectsQuery, returning a list of object IDs
-func (q *ListObjectsQuery) Execute(ctx context.Context, req *openfgapb.ListObjectsRequest) (*openfgapb.ListObjectsResponse, error) {
-	err := q.validateInput(ctx, req.StoreId, req.Type, req.AuthorizationModelId, req.Relation)
+type listObjectsRequest interface {
+	GetStoreId() string
+	GetAuthorizationModelId() string
+	GetType() string
+	GetRelation() string
+	GetUser() string
+	GetContextualTuples() *openfgapb.ContextualTupleKeys
+}
+
+func (q *ListObjectsQuery) handler(
+	ctx context.Context,
+	req listObjectsRequest,
+	resolvedChan chan<- struct{},
+	resultsChan chan<- string,
+	errChan chan<- error,
+) error {
+
+	targetObjectType := req.GetType()
+	targetRelation := req.GetRelation()
+
+	model, err := q.Datastore.ReadAuthorizationModel(ctx, req.GetStoreId(), req.GetAuthorizationModelId())
 	if err != nil {
-		return nil, err
+		if errors.Is(err, storage.ErrNotFound) {
+			return serverErrors.TypeNotFound(targetObjectType)
+		}
+		return err
+	}
+
+	typesys := typesystem.New(model)
+
+	_, ok := typesys.GetTypeDefinition(targetObjectType)
+	if !ok {
+		return serverErrors.TypeNotFound(targetObjectType)
+	}
+
+	relation, ok := typesys.GetRelation(targetObjectType, targetRelation)
+	if !ok {
+		return serverErrors.RelationNotFound(targetRelation, targetObjectType, nil)
+	}
+
+	hasTypeInfo, ok := typesys.HasTypeInfo(targetObjectType, targetRelation)
+	if !ok {
+		q.Logger.WarnWithContext(
+			ctx, fmt.Sprintf("failed to lookup type info for relation '%s'", targetRelation),
+			zap.String("storeID", req.GetStoreId()),
+			zap.String("objectType", targetObjectType),
+		)
+	}
+
+	handler := func() {
+		q.performChecks(ctx, req, resultsChan, errChan, resolvedChan)
 	}
 
+	// ConnectedObjects currently only supports models that do not include intersection and exclusion,
+	// and the model must include type info for ConnectedObjects to work.
+	if !typesystem.RewriteContainsIntersection(relation) &&
+		!typesystem.RewriteContainsExclusion(relation) &&
+		hasTypeInfo {
+
+		userObj, userRel := tuple.SplitObjectRelation(req.GetUser())
+
+		handler = func() {
+			err = q.ConnectedObjects(ctx, &ConnectedObjectsRequest{
+				StoreID:          req.GetStoreId(),
+				ObjectType:       targetObjectType,
+				Relation:         targetRelation,
+				User:             &openfgapb.ObjectRelation{Object: userObj, Relation: userRel},
+				ContextualTuples: req.GetContextualTuples().GetTupleKeys(),
+				Limit:            q.ListObjectsMaxResults,
```

Comment:
Minor: everything being passed here is related to the request, except for `q.ListObjectsMaxResults`. Maybe move `Limit` to the `ConnectedObjectsCommand` struct?

Reply:
❤️ the recommendation. I've made these changes. Much better!